File: app/src/main/java/com/example/client/screens/main/activity/MainActivity.java
package com.example.client.screens.main.activity;

import androidx.appcompat.app.AppCompatActivity;
import androidx.cardview.widget.CardView;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.example.client.R;

import com.example.client.app.Constants;
import com.example.client.models.event.Event;
import com.example.client.models.noti.Notification;
import com.example.client.models.order.OrderModel;
import com.example.client.screens.cart.activity.CartActivity;
import com.example.client.screens.home.fragment.HomeFragment;
import com.example.client.screens.main.present.MainPresent;
import com.example.client.screens.map.activity.MapsActivity;
import com.example.client.screens.noti.activity.NotificationActivity;
import com.example.client.screens.order.detail.OrderDetailActivity;
import com.example.client.screens.order.history.activity.OrderHistoryActivity;
import com.example.client.screens.profile.fragment.ProfileFragment;
import com.example.client.screens.wallet.fragment.WalletFragment;
import com.google.android.material.bottomnavigation.BottomNavigationView;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.text.MessageFormat;


public class MainActivity extends AppCompatActivity implements IMainView, View.OnClickListener{
    private BottomNavigationView navigation;
    private FragmentManager fragmentManager;
    private FragmentTransaction fragmentTransaction;
    private MainPresent mPresent;
    private ImageView imvNoti;
    private CardView cvCartPlace;
    private TextView tvQuantity;
    private RelativeLayout rllCountOrder, rllOrder, rllLoading;
    private TextView tvCountOrder, tvSeeMore, tvOrderStatus, tvOrderAddress, tvSeeOrder;
    private OrderModel order;
    private String tag = HomeFragment.class.getName();
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        imvNoti = findViewById(R.id.imv_noti);
        navigation = findViewById(R.id.navigation);
        cvCartPlace = findViewById(R.id.cv_cart_place);
        tvQuantity = findViewById(R.id.tv_quantity);
        rllCountOrder = findViewById(R.id.rll_count_order);
        rllOrder = findViewById(R.id.rll_order);
        rllLoading = findViewById(R.id.rll_loading);
        tvCountOrder = findViewById(R.id.tv_count_order);
        tvSeeMore = findViewById(R.id.tv_see_more);
        tvOrderStatus = findViewById(R.id.tv_order_status);
        tvOrderAddress = findViewById(R.id.tv_order_address);
        tvSeeOrder = findViewById(R.id.tv_see_order);

        mPresent = new MainPresent(this);


        mPresent.replaceFragment(R.id.menu_home);
        mPresent.getCartFromRes();

        mPresent.getListOrderFromService();

//        navigation.getOrCreateBadge(R.id.menu_noti).setVisible(true);
//        navigation.getOrCreateBadge(R.id.menu).setNumber(99);

        navigation.setOnItemSelectedListener(item -> {
            if(!item.isChecked()){
                switch (item.getItemId()){
                    case R.id.menu_home:
                        mPresent.replaceFragment(R.id.menu_home);
                        break;
                    case R.id.menu_wallet:
                        mPresent.replaceFragment(R.id.menu_wallet);
                        break;
                    case R.id.menu_profile:
                        mPresent.replaceFragment(R.id.menu_profile);
                        break;
                }
            }
            return true;
        });

        imvNoti.setOnClickListener(this);
        cvCartPlace.setOnClickListener(this);
        tvSeeMore.setOnClickListener(this);
        tvSeeOrder.setOnClickListener(this);
    }


    @Override
    public void showHomeScreen() {
        tag = HomeFragment.class.getName();

        fragmentManager=getSupportFragmentManager();
        fragmentTransaction=fragmentManager.beginTransaction();
        if (!fragmentManager.popBackStackImmediate (tag, 0)) {
            fragmentTransaction.add(R.id.frameLayout,new HomeFragment()).addToBackStack(tag).commit();
        } else {
            fragmentManager.popBackStack(tag,0);
        }
    }

    @Override
    public void showNotiScreen() {
        tag = WalletFragment.class.getName();

        fragmentManager=getSupportFragmentManager();
        fragmentTransaction=fragmentManager.beginTransaction();
        if (!fragmentManager.popBackStackImmediate (tag, 0)) {
            fragmentTransaction.add(R.id.frameLayout,new WalletFragment()).addToBackStack(tag).commit();
        } else {
            fragmentManager.popBackStack(tag,0);
        }
    }

    @Override
    public void showProfileScreen() {
        tag = ProfileFragment.class.getName();
        fragmentManager=getSupportFragmentManager();
        fragmentTransaction=fragmentManager.beginTransaction();
        if (!fragmentManager.popBackStackImmediate (tag, 0)) {
            fragmentTransaction.add(R.id.frameLayout,new ProfileFragment()).addToBackStack(tag).commit();
        } else {
            fragmentManager.popBackStack(tag,0);
        }
    }

    @Override
    public void onBackPressed() {
        finish();
    }

    @Override
    public void showCart(int quantity) {
        cvCartPlace.setVisibility(View.VISIBLE);
        tvQuantity.setText(MessageFormat.format("{0}", quantity));
    }

    @Override
    public void hideCart() {
        cvCartPlace.setVisibility(View.GONE);
    }

    @Override
    public void showLoading() {
        rllLoading.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideLoading() {
        rllLoading.setVisibility(View.GONE);
    }

    @Override
    public void showErrorMessage(int errMessage) {
        Toast.makeText(this, getString(errMessage), Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showOrderCount(int count) {
        rllCountOrder.setVisibility(View.VISIBLE);
        tvCountOrder.setText(getString(R.string.text_count_order).replace("%s", count + ""));
    }

    @Override
    public void hideOrderCount() {
        rllCountOrder.setVisibility(View.GONE);

    }

    @Override
    public void showOrder(OrderModel order) {
        this.order = order;
        rllOrder.setVisibility(View.VISIBLE);
        tvOrderStatus.setText(order.getStatusString());
        tvOrderAddress.setText(order.getAddress());
    }

    @Override
    public void hideOrder() {
        rllOrder.setVisibility(View.GONE);
    }


    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.imv_noti:
                startActivity(new Intent(MainActivity.this, NotificationActivity.class));
                break;
            case R.id.cv_cart_place:
                startActivity(new Intent(MainActivity.this, CartActivity.class));
                break;
            case R.id.tv_see_more:
                startActivity(new Intent(MainActivity.this, OrderHistoryActivity.class));
                break;
            case R.id.tv_see_order:
                startActivity(OrderDetailActivity.Companion.newInstance(MainActivity.this, order.getOrdercode()));
                break;
        }
    }

    @Override
    public void onStart() {
        super.onStart();
        if (!EventBus.getDefault().isRegistered(this))
            EventBus.getDefault().register(this);
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onEvent(Event event) {
        switch (event.getKey()) {
            case Constants.EventKey.UPDATE_CART:
                mPresent.getCartFromRes();
                break;
            case Constants.EventKey.UPDATE_STATUS_ORDER:
                mPresent.getListOrderFromService();
                break;

        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        EventBus.getDefault().unregister(this);
    }

}

File: app/src/main/java/com/example/client/screens/main/activity/MainActivity.kt
package com.example.client.screens.main.activity

import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import com.example.client.R
import com.example.client.base.BaseActivityMVP
import com.example.client.models.order.OrderModel
import com.example.client.screens.branch.BranchActivity
import com.example.client.screens.cart.activity.CartActivity
import com.example.client.screens.home.fragment.HomeFragment
import com.example.client.screens.main.present.IMainPresent
import com.example.client.screens.main.present.MainPresent
import com.example.client.screens.map.activity.MapsActivity
import com.example.client.screens.order.detail.OrderDetailActivity
import com.example.client.screens.order.history.activity.OrderHistoryActivity
import com.example.client.screens.payment.PaymentActivity
import com.example.client.screens.product.activity.ProductActivity
import com.example.client.screens.profile.fragment.ProfileFragment
import com.example.client.screens.wallet.fragment.WalletFragment
import com.example.client.utils.ActivityUtils
import com.example.client.utils.LocationUtils
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.gms.location.LocationServices
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : BaseActivityMVP<IMainPresent>(), IMainView, View.OnClickListener, GoogleApiClient.ConnectionCallbacks, GoogleApiClient.OnConnectionFailedListener {

    private var tag = HomeFragment::class.java.name
    private var showOrderCount = false
    private var showOrder = false
    private var showCart = false

    companion object {
        @JvmStatic
        fun newInstance(from: Activity): Intent = Intent(from, ProductActivity::class.java)
    }

    override val presenter: IMainPresent
        get() = MainPresent(this)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

//        navigation.getOrCreateBadge(R.id.menu_noti).setVisible(true);
//        navigation.getOrCreateBadge(R.id.menu).setNumber(99);
    }

    override fun bindComponent() {
        showHomeScreen()
    }

    override fun bindData() {
        requestLocationPermission()
    }

    private fun requestLocationPermission() {
        presenter.run {
            if (LocationUtils.isEnableGPS(this@MainActivity)) {
                LocationUtils.requestLocationPermission(this@MainActivity) {
                    updateCurrentLocation()
                }
            } else {
                val mGoogleApiClient = GoogleApiClient.Builder(this@MainActivity)
                        .addApi(LocationServices.API)
                        .addConnectionCallbacks(this@MainActivity)
                        .addOnConnectionFailedListener(this@MainActivity).build()
                mGoogleApiClient.run {
                    connect()
                    LocationUtils.requestGPS(this, this@MainActivity)
                }
            }
            bindData()
        }
    }


    override fun bindEvent() {
        navigation.setOnItemSelectedListener { item: MenuItem ->
            if (!item.isChecked) {
                when (item.itemId) {
                    R.id.menu_home -> showHomeScreen()
                    R.id.menu_wallet -> showWalletScreen()
                    R.id.menu_profile -> showProfileScreen()
                }
                showOrderInfo()
            }
            true
        }
        imv_notification.setOnClickListener(this)
        cv_cart_place.setOnClickListener(this)
        tv_see_more.setOnClickListener(this)
        tv_see_order.setOnClickListener(this)
    }

    private fun showHomeScreen() {
        tag = HomeFragment::class.java.simpleName
        ActivityUtils.addFragmentToActivity(supportFragmentManager, HomeFragment.newInstance(), R.id.frame_layout, tag)
    }

    private fun showWalletScreen() {
        tag = WalletFragment::class.java.simpleName
        ActivityUtils.addFragmentToActivity(supportFragmentManager, WalletFragment.newInstance(), R.id.frame_layout, tag)
    }

    private fun showProfileScreen() {
        tag = ProfileFragment::class.java.simpleName
        ActivityUtils.addFragmentToActivity(supportFragmentManager, ProfileFragment.newInstance(), R.id.frame_layout, tag)
    }

    fun showOrderInfo() {
        if (tag == HomeFragment::class.java.simpleName) {
            cv_cart_place.visibility = if (showCart) View.VISIBLE else View.GONE
            rll_count_order.visibility = if (showOrderCount) View.VISIBLE else View.GONE
            rll_order.visibility = if (showOrder) View.VISIBLE else View.GONE
        } else {
            cv_cart_place.visibility = View.GONE
            rll_count_order.visibility = View.GONE
            rll_order.visibility = View.GONE
        }
    }

    override fun showCart(quantity: Int) {
        showCart = true
        cv_cart_place.visibility = View.VISIBLE
        tv_cart_quantity.text = quantity.toString()
    }

    override fun hideCart() {
        showCart = false
        cv_cart_place.visibility = View.GONE
    }

    override fun showLoading() {
        rll_loading.visibility = View.VISIBLE
    }

    override fun hideLoading() {
        rll_loading.visibility = View.GONE
    }

    override fun onBackPress() {}

    override fun onBackPressed() {
        finish()
    }

    override fun showErrorMessage(errMessage: Int) {
        showToastMessage(getString(errMessage))
    }

    override fun showOrderCount(count: Int) {
        showOrderCount = true
        rll_count_order.visibility = View.VISIBLE
        tv_count_order.text = getString(R.string.text_count_order, count)
    }

    override fun hideOrderCount() {
        showOrderCount = false
        rll_count_order.visibility = View.GONE
    }

    override fun showOrder(order: OrderModel) {
        showOrder = true
        rll_order.visibility = View.VISIBLE
        tv_order_description.text = order.getStatusString().plus("\n").plus(getString(R.string.main_delivery_to)).plus(order.address)
    }

    override fun hideOrder() {
        showOrder = false
        rll_order.visibility = View.GONE
    }

    override fun navigateToOrderDetail(orderCode: String) {
        startActivity(OrderDetailActivity.newInstance(this@MainActivity, orderCode))
    }

    override fun toBranchScreen() {
        startActivity(Intent(this, BranchActivity::class.java))
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.imv_notification -> startActivity(Intent(this, PaymentActivity::class.java))
            R.id.cv_cart_place -> startActivity(Intent(this, CartActivity::class.java))
            R.id.tv_see_more -> startActivity(OrderHistoryActivity.newInstance(this))
            R.id.tv_see_order -> presenter.navigateToOrderDetail()
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        when (requestCode) {
            MapsActivity.PERMISSIONS_REQUEST_ACCESS_FINE_LOCATION -> {
                if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    presenter.updateCurrentLocation()
                } else showErrorMessage(R.string.GPS_not_found)
            }
            else -> super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        }
    }

    override fun onConnected(p0: Bundle?) {
    }

    override fun onConnectionSuspended(p0: Int) {
    }

    override fun onConnectionFailed(p0: ConnectionResult) {
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        when (requestCode) {
            LocationUtils.REQUEST_LOCATION -> when (resultCode) {
                RESULT_OK -> {
                    requestLocationPermission()
                }
                else -> showErrorMessage(R.string.GPS_not_found)
            }
        }
    }

}